### ISSUE TITLE
[bug] ifnone state-dependent path delay

### ISSUE THREAD

**Issue by likeamahoney** [IHP-GmbH/IHP-Open-PDK#209]
Hi, All!

I found that `ifnone` condition used [here](https://github.com/IHP-GmbH/IHP-Open-PDK/blob/5a42d03194e8c98558f4e34538338a60550f89b9/ihp-sg13g2/libs.ref/sg13g2_stdcell/verilog/sg13g2_stdcell.v#L1562) and [here](https://github.com/IHP-GmbH/IHP-Open-PDK/blob/5a42d03194e8c98558f4e34538338a60550f89b9/ihp-sg13g2/libs.ref/sg13g2_stdcell/verilog/sg13g2_stdcell.v#L1567) is illegal due to SystemVerilog LRM IEEE 1800-2017 and it's previous versions due to section 30.4.4.4 - The ifnone condition description:

> The ifnone keyword is used to specify a default state-dependent path delay when all other conditions for
> the path are false. The ifnone condition shall specify the same module path source and destination as the
> state-dependent module paths. The following rules apply to module paths specified with the ifnone
> condition:
> —
> Only simple module paths may be described with an ifnone condition.
> —The state-dependent paths that correspond to the ifnone path may be either simple module paths or
> edge-sensitive paths

> —If there are no corresponding state-dependent module paths to the ifnone module path, then the
> ifnone module path shall be treated the same as an unconditional simple module path.
> —
> It is illegal to specify both an ifnone condition for a module path and an unconditional simple
> module path for the same module path.
> 

Which means that edge-sensitive path can't be specified under `ifnone` condition. Edge sensitive paths can be specified only in `if` conditions which correspond to `ifnone` .

**Comment by Blebowski** [comment]
Hi @likeamahoney,

maybe I got it wrong, but doesn't the following:
```
The state-dependent paths that correspond to the ifnone path may be either simple module paths or
**edge-sensitive paths**
```
allow having the edge-sensitive paths ?

**Comment by sergeiandreyev** [comment]
Hi @Blebowski, there was a thread on this issue:
https://sourceforge.net/p/iverilog/feature-requests/32

**Comment by Blebowski** [comment]
Oh, I see, thanks.

It seems like inconsistency in the LRM itself, or not ?

The grammar says:
```
state_dependent_path_declaration ::=
      if ( module_path_expression ) simple_path_declaration
   | if ( module_path_expression ) edge_sensitive_path_declaration
   | ifnone simple_path_declaration
```

but the description says:

```
The state-dependent paths that correspond to the ifnone path may be either simple module paths or
edge-sensitive paths.
```

which would imply that edge-sensitive paths should be allowed here.

**Comment by Blebowski** [comment]
Oh, I see now:
```
You are misinterpreting B. It says the paths that correspond to the ifnone may be either simple or edge-sensitive not that the ifnone path can be edge-sensitive. For this definition correspond means the other state-dependent delays that have the same input and output terminals as the ifnone.

FYI this text is basically straight from the standard. 
```

**Comment by sergeiandreyev** [comment]
yes, basically we need to figure out how to update our `specify` blocks in Verilog models to be compatible with open source tools (which strictly follow the standard)
we did not catch the issue, because we're using commercial tools for PDK validation, and they have some features on top of the standard as it shows up

**Comment by b10346** [comment]
It is solved in the latest standard cell update or not actual at the moment.
Could anybody check the status and confirm that the issue can be closed?

**Comment by sergeiandreyev** [comment]
> It is solved in the latest standard cell update or not actual at the moment.
Could anybody check the status and confirm that the issue can be closed?

Hi @likeamahoney, is this issue still valid? (see above)

**Comment by likeamahoney** [comment]
> It is solved in the latest standard cell update or not actual at the moment. Could anybody check the status and confirm that the issue can be closed?

> > It is solved in the latest standard cell update or not actual at the moment.
> > Could anybody check the status and confirm that the issue can be closed?
> 
> Hi [@likeamahoney](https://github.com/likeamahoney), is this issue still valid? (see above)

The issue is not resolved.

There are now even more problematic places in the standard cells, for example [here](https://github.com/IHP-GmbH/IHP-Open-PDK/blob/331c00484213b13414777eec1336ef5c29b969bd/ihp-sg13g2/libs.ref/sg13g2_stdcell/verilog/sg13g2_stdcell.v#L42) and [here](https://github.com/IHP-GmbH/IHP-Open-PDK/blob/331c00484213b13414777eec1336ef5c29b969bd/ihp-sg13g2/libs.ref/sg13g2_stdcell/verilog/sg13g2_stdcell.v#L44) (and throughout the rest of the code).

As mentioned by that user, edge-sensitive cells cannot be used in ifnone according to the LRM:

> 30.4.4.4 The ifnone condition
> The ifnone keyword is used to specify a default state-dependent path delay when all other conditions for
> the path are false. The ifnone condition shall specify the same module path source and destination as the
> state-dependent module paths. The following rules apply to module paths specified with the ifnone
> condition:
> — Only simple module paths may be described with an ifnone condition.
> — The state-dependent paths that correspond to the ifnone path may be either simple module paths or
> edge-sensitive paths.
> — If there are no corresponding state-dependent module paths to the ifnone module path, then the
> ifnone module path shall be treated the same as an unconditional simple module path.
> — It is illegal to specify both an ifnone condition for a module path and an unconditional simple
> module path for the same module path.

I also tested this with commercial tools (like VCS and Questa), and unlike the open-source tools, they do accept this construct. However, this still appears to be non-compliant with the LRM.

If supporting the behavior accepted by commercial tools is sufficient for your purposes, then I suppose this issue can be closed. Otherwise, these cases should still be addressed.
